reader_concurrency_semaphore.cc

```diff
@@ -993,6 +993,12 @@ future<> reader_concurrency_semaphore::with_ready_permit(reader_permit permit, r
     return fut;
 }
 
+void reader_concurrency_semaphore::set_resources(resources r) {
+    auto delta = r - _initial_resources;
+    _initial_resources = r;
+    _resources += delta;
```

Comment:
This evaluates to _resource = _resources + r - _initial_resources. I don't understand it.

Reply:
Ah, after thinking I get it.

Reply:
moral: don't think

Reply:
Will things continue to work if _resources go negative? I think they will.

Reply:
Looking at reader_resources:

```
    explicit operator bool() const {
        return count > 0 || memory > 0;
    }
```

Should that || be &&?

Reply:
> Will things continue to work if _resources go negative? I think they will.

Yes. Memory can routinely go into negative. Count should never go negative though. It can briefly do after a `set_resources()` call, but it will go >= 0 after that on its own.

Reply:
> Looking at reader_resources:
> 
> ```
>     explicit operator bool() const {
>         return count > 0 || memory > 0;
>     }
> ```
> 
> Should that || be &&?

I remember thinking on that a lot when writing it. I think `&&` breaks the current code using `operator bool()`.

Reply:
I think that it's wrong to have operator bool here, or operator>=. Mathematically this is a vector and vectors aren't ordered like scalars.

You can see it by trying to synthesize operator< from operator>= and then arrive at a contraction.

Current operator>=:

```c++
    bool operator>=(const reader_resources& other) const {
        return count >= other.count && memory >= other.memory;
    }
```

"reasonable" operator>:

```c++
    bool operator>(const reader_resources& other) const {
        return count > other.count && memory > other.memory;
    }
```

Swapping the order of arguments to get operator<:

```c++
    bool operator<(const reader_resources& other) const {
        return count < other.count && memory < other.memory;
    }
```

Synthesizing operator< from !operator>=:

```c++
    bool operator<(const reader_resources& other) const {
        return !operator>=(other);
    }
```

Manually inlining:

```c++
    bool operator<(const reader_resources& other) const {
        return !(count >= other.count && memory >= other.memory);
    }
```

Applying de-Morgan and converting ! >= to <:

```c++
    bool operator<(const reader_resources& other) const {
        return count < other.count || memory < other.memory;
    }
```

So we arrived at two different definitions of operator<, one using && and one using ||.

Reply:
I'll look into removing these and inlining them into their callers. They are indeed problematic.

Reply:
I'll look into removing these and inlining them into their callers. They are indeed problematic.

Reply:
https://github.com/scylladb/scylladb/pull/11813